### DISCUSSION TITLE
M2M joining table with multiple same-table FKs?

### DISCUSSION THREAD

**Discussion by wmshort** [piccolo-orm/piccolo#388] [Q&A]
@dantownsend I have been playing around with the excellent new M2M functionality and API; really glad to see that this has been added. I have a couple of comments and questions. First, I understand the reasoning for 'module_path' but I wonder if there isn't a way to hide this or only expose it when absolutely necessary. I don't find it intuitive and it seems to show a bit too much of the 'guts' of the mechanism to novice users. Second, perhaps more immediate: is there some mechanism for specifying the field on the joining table to which a M2M field 'resolves'? What I mean is that I can imagine cases where a joining table has two fields referencing the same other table; let's say something like:

class Band(Table):
     opponents = M2M(LazyTableReference("BattleOfTheBands"...))

class BattleOfTheBands(Table): 
     band1 = ForeignKey(references=Band)
     band2 = ForeignKey(references=Band)

Is there a way of 'pointing' the M2M declaration at one of the fields on the joining table (e.g. band1), or am I simply thinking about this wrong?

**Comment by dantownsend** [comment]
@wmshort I've been thinking the same thing. As you say, if an API like `LazyTableReference("BattleOfTheBands")` existed, it would be much nicer.

The reason I didn't do it originally was because I had to work out which ​module `LazyTableReference` was instantiated in.

Finding this out is quite tricky - if you use `self.__module__` it tells you where the `LazyTableReference` class was declared, not where it was instantiated. I came across suggestions on StackOverflow, which use the `inspect` module to look at the stack:

https://stackoverflow.com/a/22147515

This feels a bit hacky though.

The best solution seems to be using the `Table` Metaclass. It knows which module the `Table` subclass was declared in, so by extension I know which module the `LazyTableReference` was instantiated in. I might give this a quick go and see if it works.

**Comment by dantownsend** [comment]
@wmshort With the second point, there isn't a way of doing this currently. Piccolo works out which foreign key you're interested in on the joining table automatically (i.e. does this foreign key point at me). I didn't consider the use case of a joining table where both foreign keys reference the same table. Supporting this shouldn't be too hard though.
